### PR TITLE
Reduce memory allocations for Process.GetProcessesByName

### LINKED ISSUE #40768: Can the GetProcessesByName method reduce the number of arrays and Process objects created?
### Description

As written in the code, GetProcessesByName first calls GetProcesses to obtain all processes of the machine, and then filters the process name

```
        public static Process[] GetProcessesByName(string? processName, string machineName)
        {
            if (processName == null)
            {
                processName = string.Empty;
            }

            Process[] procs = GetProcesses(machineName);
            var list = new List<Process>();

            for (int i = 0; i < procs.Length; i++)
            {
                if (string.Equals(processName, procs[i].ProcessName, StringComparison.OrdinalIgnoreCase))
                {
                    list.Add(procs[i]);
                }
                else
                {
                    procs[i].Dispose();
                }
            }

            return list.ToArray();
        }
```

But in fact, we can filter the process name in GetProcesses in advance.

```
        public static Process[] GetProcesses(string machineName)
        {
            bool isRemoteMachine = ProcessManager.IsRemoteMachine(machineName);
            ProcessInfo[] processInfos = ProcessManager.GetProcessInfos(machineName);
            Process[] processes = new Process[processInfos.Length];
            for (int i = 0; i < processInfos.Length; i++)
            {
                ProcessInfo processInfo = processInfos[i];
                processes[i] = new Process(machineName, isRemoteMachine, processInfo.ProcessId, processInfo);
            }
            return processes;
        }
```

If we do this, we can reduce the allocation length of the Process array and create some Process objects in the GetProcesses method

### PR DESCRIPTION
**Pull request by Serg046**
Fix #40768

Seems felipepessoto's suggestion is not the best choice because there is no any issue on Linux and we have different code paths for local/remote machines on Windows.

### CODE REVIEW

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs [hunk 1]
```diff
@@ -28,22 +28,8 @@ public static Process[] GetProcessesByName(string? processName, string machineNa
...
-
-            for (int i = 0; i < procs.Length; i++)
-            {
-                if (string.Equals(processName, procs[i].ProcessName, StringComparison.OrdinalIgnoreCase))
-                {
-                    list.Add(procs[i]);
-                }
-                else
-                {
-                    procs[i].Dispose();
```
Comment (lkts) [inline src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs:42]: this `Dispose` is missing in new version, isn't it?

**Comment by adamsitnik** [comment]
> It would be useful to benchmark both time and total allocations for baseline and different variants of the fix.

I agree with @jkotas, we need benchmark results to see how this change is affecting performance

@Serg046 could you please follow the instructions from https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md#Preventing-Regressions and run [these benchmarks](https://github.com/dotnet/performance/blob/master/src/benchmarks/micro/libraries/System.Diagnostics/Perf_Process.cs) against `CoreRun.exe` with and withour your changes?

**Comment by Serg046** [comment]
@adamsitnik, here we see less allocations but when I ran it next time the number of allocations was almost the same. If we really want to prove that the change is beneficial, we will need a more accurate benchmark with more stable data than the processes and their related code branches. Also please let me know if we want to go deeper in order to further reduce allocations.

Before:
|                     Method |        Job | UnrollFactor |            Mean |           Error |          StdDev |          Median |              Min |             Max |   Gen 0 |   Gen 1 | Gen 2 | Allocated |
|--------------------------- |----------- |------------- |----------------:|----------------:|----------------:|----------------:|-----------------:|----------------:|--------:|--------:|------:|----------:|
|               GetProcesses | Job-WOZLYI |           16 |  2,637,525.2 ns |     9,577.38 ns |     7,997.55 ns |  2,636,506.2 ns |  2,624,027.08 ns |  2,648,383.3 ns | 72.9167 | 20.8333 |     - |  392533 B |
|         GetProcessesByName | Job-WOZLYI |           16 |  2,642,911.2 ns |    13,237.25 ns |    12,382.14 ns |  2,640,937.0 ns |  2,620,994.27 ns |  2,668,576.6 ns | 72.9167 | 20.8333 |     - |  392952 B |

After:
|                     Method |        Job | UnrollFactor |            Mean |           Error |          StdDev |          Median |              Min |             Max |   Gen 0 |   Gen 1 | Gen 2 | Allocated |
|--------------------------- |----------- |------------- |----------------:|----------------:|----------------:|----------------:|-----------------:|----------------:|--------:|--------:|------:|----------:|
|               GetProcesses | Job-SBOBLH |           16 |  2,585,315.5 ns |    14,779.43 ns |    12,341.50 ns |  2,586,216.5 ns |  2,565,340.21 ns |  2,600,253.6 ns | 72.1649 | 30.9278 |     - |  384679 B |
|         GetProcessesByName | Job-SBOBLH |           16 |  2,549,940.4 ns |    36,529.20 ns |    34,169.44 ns |  2,530,560.7 ns |  2,514,150.00 ns |  2,608,208.9 ns | 62.5000 | 17.8571 |     - |  304584 B |

Diff:
| Slower                                                            | diff/base | Base Median (ns) | Diff Median (ns) | Modality|
| ----------------------------------------------------------------- | ---------:| ----------------:| ----------------:| --------:|
| System.Diagnostics.Perf_Activity.EnumerateActivityTagsSmall       |      1.15 |            51.06 |            58.81 |         |
| System.Diagnostics.Perf_Activity.EnumerateActivityTagsLarge       |      1.09 |          8636.64 |          9384.71 |         |
| System.Diagnostics.Perf_Activity.EnumerateActivityTagObjectsLarge |      1.04 |         15011.17 |         15636.22 |         |
| System.Diagnostics.Perf_Activity.EnumerateActivityLinkTagsLarge   |      1.03 |         14204.02 |         14696.13 |         |

| Faster                                                        | base/diff | Base Median (ns) | Diff Median (ns) | Modality|
| ------------------------------------------------------------- | ---------:| ----------------:| ----------------:| --------:|
| System.Diagnostics.Perf_Process.GetProcessesByName            |      1.04 |       2640936.98 |       2530560.71 |         |
| System.Diagnostics.Perf_Activity.EnumerateActivityEventsSmall |      1.03 |            96.88 |            94.48 |         |

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.cs [hunk 2]
```diff
@@ -1054,14 +1054,36 @@ public static Process[] GetProcesses()
...
         {
             bool isRemoteMachine = ProcessManager.IsRemoteMachine(machineName);
             ProcessInfo[] processInfos = ProcessManager.GetProcessInfos(machineName);
             Process[] processes = new Process[processInfos.Length];
+            int processesLength = 0;
             for (int i = 0; i < processInfos.Length; i++)
             {
                 ProcessInfo processInfo = processInfos[i];
-                processes[i] = new Process(machineName, isRemoteMachine, processInfo.ProcessId, processInfo);
+                if (processFilter == null || processFilter(processInfo))
```
[the remark this answers is not in the document]
Comment (iSazonov) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.cs:1077]: > I thinks it's better to go a step back and eleminate the allocations from ProcessManager.GetProcessInfos.

We could use the same approach as File/Directory enumeration API does with `public ref partial struct FileSystemEntry`.

**Comment by danmosemsft** [comment]
@adamsitnik should we close this (at least for now)?

**Comment by stephentoub** [comment]
> @adamsitnik should we close this (at least for now)?

@adamsitnik?

**Comment by danmosemsft** [comment]
I'm going to close this Pr as it is stalled. Thank you for the submission though.

**Pull request closed without merging**